Sources/Helpers/CellItemFactory.swift

```diff
@@ -0,0 +1,254 @@
+        if let array = array as? [U] {
+            return makeCellItems(for: array)
+        }
+        return []
+    }
+    
+    public func makeCellItems(for object: Any, at index: Int) -> [CollectionViewCellItem] {
+        if let object = object as? U {
+            if let initializationHandler = self.initializationHandler {
+                var cellItems = [CollectionViewCellItem]()
+                initializationHandler(index, object).forEach { cellItem in
+                    if let cellItem = cellItem {
+                        cellItems.append(cellItem)
+                    }
+                }
+                return cellItems
+            }
+            else {
+                return [makeUniversalCellItem(for: object, at: index)]
+            }
+        }
+        return []
+    }
+    
+    public func join(_ factory: CellItemFactory) -> CellItemFactory {
+        let complexFactory = ComplexCellItemFactory()
+        complexFactory.join(self)
+        complexFactory.join(factory)
+        return complexFactory
+    }
+    
+    public var hashKey: String? {
+        return String(describing: U.self)
+    }
+}
+
+public class ComplexCellItemFactory: CellItemFactory {
+    
+    private var factories = [String: CellItemFactory]()
+    
+    public init() {
+    }
+
+    public func makeCellItems(for array: [Any]) -> [CollectionViewCellItem] {
+        var cellItems = [CollectionViewCellItem]()
+        array.enumerated().forEach { index, object in
+            if let factory = factories[String(describing: type(of: object))] {
+                cellItems.append(contentsOf: factory.makeCellItems(for: object, at: index))
+            }
+        }
+        return cellItems
+    }
+    
+    public func makeCellItems(for object: Any, at index: Int) -> [CollectionViewCellItem] {
+        if let factory = factories[String(describing: type(of: object))] {
+            return factory.makeCellItems(for: object, at: index)
+        }
+        return []
+    }
+    
+    @discardableResult
+    public func join(_ factory: CellItemFactory) -> CellItemFactory {
+        if let factory = factory as? ComplexCellItemFactory {
+            factory.factories.forEach { (key, value) in
+                factories[key] = value
+            }
+        }
+        else if let key = factory.hashKey {
+            factories[key] = factory
+        }
+        return self
+    }
+    
+    /// Removes a factory from complex hierarchy
+    ///
+    /// - Parameters:
+    ///     - factory: a factory that should be removed
+    public func unjoin(_ factory: CellItemFactory) {
+        if let key = factory.hashKey {
+            factories.removeValue(forKey: key)
+        }
+        else if let factory = factory as? ComplexCellItemFactory {
+            factory.factories.forEach { (key, _) in
+                factories.removeValue(forKey: key)
+            }
+        }
+    }
+    
+    public var hashKey: String? {
+        return nil
+    }
```

Comment:
Is it normal to use `nil` hashKey? Or it should be required?

Reply:
yes, it means that hash association is ambiguous